articles/node_lifecycle.md

```diff
@@ -0,0 +1,176 @@
+
+Any service requests to a node in the inactive state will not be answered (to the caller, they will fail immediately).
+
+Entering the `inactive` state from the `created` state requires running the node's `onInitialised` method. The node uses this to set up any resources it must hold throughout its life (irrespective of if it is active or inactive). As examples, such resources may include topic publications and subscriptions, memory that is held continuously, and initialising configuration parameters.
+
+Entering the `inactive` state from the `active` state requires running the node's `onDeactivated` method. See the `active` state description for more information.
+
+Entering the `inactive` state from the `fatal error` state requires running the node's `onReset` method. See the `fatal error` state description for more information.
+
+If an error occurs while in this state that cannot be handled by the node, then the transition to the `fatal error` state is taken.
+
+The node may also exit this state by being destroyed, in which case the `destroy` transition is taken.
+
+Implementation note: Causing the transition to the destroyed state can often be mapped to deletion of the node instance.
+
+### Active state
+
+This is the main state of the node's life cycle. While in this state, the node performs any processing, responds to service requests, reads and processes data, produces output, etc.
+
+Entering the `active` state requires running the node's `onActivated` method. This is used to prepare the node for execution. This may include acquiring resources that are only held while the node is actually active, such as access to hardware. Ideally, no preparation that requires significant time (such as lengthy hardware initialisation) should be performed in this callback.
+
+To shut down a node, it must first be deactivated. This will cause the `onDeactivate` method to be called, allowing the node to clean up any resources it only holds while active.
+
+If an error that cannot be handled by the node/system occurs in this state, the transition to the `fatal error` state is taken.
+
+### Fatal Error state
+
+Known errors that can occur in a node should be considered designed behaviour of the system and so should be handled by the node where the error occurred and, if necessary, the surrounding nodes. The reason for this is that the system knows best how to handle errors that occur in its behaviour.
+
+Any fatal errors (those which the node/system was not expecting and does not know how to handle) will cause a transition to the `fatal error` state. In this state, because the node does not know what to do in the presence of the error, it is not provided any execution time. Arriving data is not processed, service requests are not responded to, and no output is produced.
+
+On entering the `fatal error` state, the node's `onError` method must be executed. This provides the node with a chance to clean up resources it was using.
```

Comment:
> "Fatal errors" vs "Normal errors": see my comment above.

I don't see where you mentioned "fatal errors" vs "normal errors" above, can you reiterate that or point it out directly?

> But basically when in "fatal" the only safe recovery mechanism is to tear down the process, not only the node. Since the cleanup procedures failed, you really don't know in what state you are. In C++, this includes not being very sure of your pointer's states ...

I suppose it depends on your definition of a fatal error, but depending on the nature of the error, it would not necessarily follow that the whole process must come down. From my perspective any unhandled error that a node has internally is fatal to that node, but that error could be trivial like an OS Error because a file it tried to open doesn't exist. In that case it would perhaps be appropriate to just restart the node. Tearing down the whole process when one node misbehaves is not strictly necessary in some cases and can be avoided. In fact this is a complaint about how the NodeletManager works, if an exception is raised by one nodelet then it is uncaught by the manager and goes directly to the abort handler, taking the whole process down, which is in my opinion unnecessary.

The only error I can think of that should bring the whole process down is a segmentation fault (bus error). Even something like failing to allocate memory might be solved by tearing down and freeing the node in question and restarting it. This might not be desirable behavior all the time, but it might be appropriate some times. This is something that should probably be configurable in the containing process.